### ISSUE TITLE
More specific JSON parse reviver argument in lib.d.ts

### ISSUE THREAD

**Issue by avonwyss** [Microsoft/TypeScript#6955]
Currently, the JSON interface in lib.d.ts declares the following:

```
parse(text: string, reviver?: (key: any, value: any) => any): any;
```

However, my understanding is that the spec guaratees a string for the key (see http://www.ecma-international.org/ecma-262/5.1/#sec-15.12.2 paragraph "The abstract operation Walk ... String name ..."), in consequence it could be more specific:

```
parse(text: string, reviver?: (key: string, value: any) => any): any;
```

**Comment by mhegazy** [comment]
The problem here is similar to [for..in](https://github.com/Microsoft/TypeScript/issues/6346) issues, see the fix in TS 1.8 in https://github.com/Microsoft/TypeScript/pull/6379. 

The value is a string, but it could be a numeric string in cases of array. it is not a number value (i.e. can not call toExponent on it) but it can still be used to index into an array (since at the end all index operations first do a toString in the key anyways). TypeScript, tries to strongly type index operations using index signatures. there are two, string and numeric. and if you index into an object using a type that it does not have an index signature for, you get an any. ideally, this 

so:

```ts
var list = [0, 1];
JSON.parse("[3, 4]", (k:string, v) => {
    var e = list[k]; // e is any, as list has no string indexer,
                          // also an error with --noImplicitAny
})

JSON.parse("[3, 4]", (k:number, v) => {
    var e = list[k]; // e number, as expected
    k.toExponential() // runtime error, k is not a number
})
```

in short, this was intentional. but maybe we should reconsider.

**Comment by avonwyss** [comment]
I understand your reasoning, but how is this different from the stringify replacer which is kind of a sibling to the reviver?

    stringify(value: any, replacer: (key: string, value: any) => any): string;

I see this as a callback with clearly defined arguments, the first being a string. Nothing prevents you from passing in a function which accepts (any, any), but it would prevent you from passing in (number, any) which is currently allowed but wrong to my understanding:

    function B(name: number, value: any): any {
    	...
    } 

    JSON.parse("", B); // no error as of now!

**Comment by RyanCavanaugh** [comment]
PR welcomed on this one

**Comment by austin94** [comment]
If no one has started on this yet, I can make a PR to add this.

**Comment by AnyhowStep** [comment]
Hey,

I've been wanting to get more involved with the TypeScript code base but simply lack the knowledge about the current implementation (code base is huge and intimidating, I've only written toy interpreters).

This issue looks simple enough, I think. Just seems to be updating `.d.ts` files?

-----

I looked at DefinitelyTyped/DefinitelyTyped#15744 and their `Reviver` type,

```ts
export type Reviver = (this: {}, key: string, value: any) => any;
```

I think the `this` parameter should be of type `any` or the following code will fail in strict mode,

```ts
JSON.parse(`{ "x":42 }`, function (k, v) {
    //Element implicitly has an 'any' type because type '{}' has no index signature.
    const x = this[k];
});
```
If we use `this : any`, then `x` will be of type `any` in the above code.

-----

It also looks like `JSON.stringify()` needs to be updated? The `replacer?` parameter looks like it could benefit from having the same signature as `Reviver`

**Comment by AnyhowStep** [comment]
![image](https://user-images.githubusercontent.com/5655961/47770133-9b264280-dcb4-11e8-8f8f-a95b7810f689.png)

I also just noticed there are trailing spaces in 4 lines of comments in `lib.es5.d.ts`.

I assume no one wants these removed.

-----

I made the changes in the two places and `git diff` gives me this,

![image](https://user-images.githubusercontent.com/5655961/47770279-1daf0200-dcb5-11e8-8c72-3018c0c4f21f.png)

-----

Assuming this is all good, do I just need to run `jake tests`, and `jake runtests` before submitting a PR?

-----

Also, I tried signing the CLA (Contributor License Agreement) at https://cla.opensource.microsoft.com/ but I think it's broken.
I signed in to GitHub and authorized the site to access my public account info and it's still asking me to sign in. Clicking "Sign In" again doesn't ask me to authorize the site again; so I assume authorization has been given but something else is broken.

Unless signing in was the same as signing the CLA...
